### PR TITLE
Fixed units-extension bugs and Units ATF reference

### PR DESCRIPTION
**Pull request by jyluo**
- Fixed a bug where units checker extension projects' UnitsRelations classes were not being loaded properly
- Updated units-extension demo project with correct implementation of FrequencyRelations, added an additional test case to ensure that scalar/second results in @Hz annotation
- Fixed reference issues to the set of units relations in Units ATF

### CODE REVIEW

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -48,7 +48,7 @@
 
     // Map from canonical class name to the corresponding UnitsRelations instance.
     // We use the string to prevent instantiating the UnitsRelations multiple times.
-    private Map<String, UnitsRelations> unitsRel;
+    private static Map<String, UnitsRelations> unitsRel;
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: Don't make things static. Each checker/atf should manage their state in instance fields.

Comment (jyluo) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: Commit 463c603 below fixed it.

File: checker/src/org/checkerframework/checker/units/UnitsTools.java [hunk 2]
```diff
@@ -27,21 +29,31 @@
 
 // TODO: add fromTo methods for all useful unit combinations.
 public class UnitsTools {
```
Comment (jyluo) [inline checker/src/org/checkerframework/checker/units/UnitsTools.java]: The static final fields are now sorted alphabetically by category to make it easier to maintain in the future.

**Comment by mernst** [comment]
There are merge conflicts in this pull request; could you resolve them?  Thanks!

**Comment by jyluo** [comment]
Fixed!

File: checker/examples/units-extension/FrequencyRelations.java [hunk 3]
```diff
@@ -35,11 +39,19 @@ public UnitsRelations init(ProcessingEnvironment env) {
...
         return null;
     }
 
-}
+    // returns true if the type mirror either has no annotations or has only 1 annotation and it is UnknownUnits
+    // any number that isn't assigned a unit will automatically gets the annotation UnknownUnits, eg: int x = 5; //x has @UnknownUnits
+    private boolean noUnits(AnnotatedTypeMirror t) {
+        Set<AnnotationMirror> annos = t.getAnnotations();
+        return annos.isEmpty() ||
+                (annos.size() == 1 &&
```
Comment (wmdietl) [inline checker/examples/units-extension/FrequencyRelations.java]: Don't you have the guarantee that you always have at least one annotation?
Why is annos.isEmpty a possibility?

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java [hunk 4]
```diff
@@ -62,7 +62,7 @@ public UnitsAnnotatedTypeFactory(BaseTypeChecker checker) {
         addTypeNameImplicit(java.lang.Void.class, BOTTOM);
     }
 
-    protected Map<String, UnitsRelations> getUnitsRel() {
+    private Map<String, UnitsRelations> getUnitsRel() {
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: Why make the method private? Is there any harm in letting a subclass use/override this method?

File: checker/examples/units-extension/Makefile [hunk 5]
```diff
@@ -10,5 +10,5 @@ compile: $(FILES)
 	$(JAVAC) $(FILES)
```
Comment (wmdietl) [inline checker/examples/units-extension/Makefile]: It would make sense to add either this example or a variant of it to the normal test suite - to ensure we don't miss issues in extensions again.

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java [hunk 6]
```diff
@@ -95,6 +94,9 @@ public AnnotationMirror aliasedAnnotation(AnnotationMirror a) {
     @Override
     @SuppressWarnings("unchecked")
     protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        // make sure we have an empty set of units relations at the start of this method
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: Why is this call of getUnitsRel necessary?

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java [hunk 7]
```diff
@@ -189,18 +190,19 @@ public AnnotationMirror aliasedAnnotation(AnnotationMirror a) {
...
                 @SuppressWarnings("unchecked")
-                Class<? extends UnitsRelations> theclass = (Class<? extends UnitsRelations>)
-                    AnnotationUtils.getElementValueClass(ama, "value", true);
+                Class<? extends UnitsRelations> theclass = (Class<? extends UnitsRelations>) AnnotationUtils.getElementValueClass(ama, "value", true);
                 String classname = theclass.getCanonicalName();
 
                 if (!getUnitsRel().containsKey(classname)) {
                     try {
-                        unitsRel.put(classname, ((UnitsRelations) theclass.newInstance()).init(processingEnv));
+                        getUnitsRel().put(classname, ((UnitsRelations) theclass.newInstance()).init(processingEnv));
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: Why is this call of getUnitsRel necessary? Two lines above it is already called, so we're sure the field was initialized.

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java [hunk 8]
```diff
@@ -103,14 +105,13 @@ public AnnotationMirror aliasedAnnotation(AnnotationMirror a) {
         } else {
             for (String qualName : qualNames.split(",")) {
                 try {
-                    final Class<? extends Annotation> q =
-                            (Class<? extends Annotation>) Class.forName(qualName);
-
+                    final Class<? extends Annotation> q = (Class<? extends Annotation>) Class.forName(qualName);
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: Can you go through your changeset and undo unnecessary whitespace changes?

File: checker/tests/units/PolyUnitTest.java [hunk 9]
```diff
@@ -4,34 +4,33 @@
 
 public class PolyUnitTest {
 
-  @PolyUnit int triplePolyUnit(@PolyUnit int amount) {
-    return 3*amount;
-  }
+    private @PolyUnit int triplePolyUnit(@PolyUnit int amount) {
```
Comment (wmdietl) [inline checker/tests/units/PolyUnitTest.java]: Visibility modifiers just add noise to test cases, you should remove them again. The other whitespace changes should also be undone. (If you feel strongly about the whitespace, make a separate pull request just for that.)

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java [hunk 10]
```diff
@@ -189,18 +190,19 @@ public AnnotationMirror aliasedAnnotation(AnnotationMirror a) {
...
      */
     private void addUnitsRelations(Class<? extends Annotation> qual) {
+        // get the annotation mirror
         AnnotationMirror am = AnnotationUtils.fromClass(elements, qual);
-
+        // get all the meta annotations of the class
         for (AnnotationMirror ama : am.getAnnotationType().asElement().getAnnotationMirrors() ) {
-            if (ama.getAnnotationType().toString().equals(UnitsRelations.class.getCanonicalName())) {
+            // if there's a UnitsRelations meta annotation
+            if (ama.getAnnotationType().toString().equals(org.checkerframework.checker.units.qual.UnitsRelations.class.getCanonicalName())) {
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: I think this was the most important change in this pull request. You can make this even clearer by extracting the UnitsRelations into a Class<? extends Annotation> variable and then comparing.

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java (hunk 6 above)
Comment (jyluo) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: I had it there at one point because unitsRel set wasn't instantiated, turns out it is removable now.

**Comment by jyluo** [comment]
I've addressed all of the code feedback in the latest commit above.

File: checker/build.xml [hunk 11]
```diff
@@ -550,6 +550,8 @@
               </fileset>
           </batchtest>
       </junit>
+
+      <antcall target="units_extension-tests"/>
```
Comment (wmdietl) [inline checker/build.xml]: Don't use underscores in the new targets (also below). All other targets use dash as separator.

File: checker/examples/units-extension/Demo.java [hunk 12]
```diff
@@ -29,4 +37,12 @@ void auto(@s int time) {
         frq = 99 / time;
     }
 
+    public static void main(String[] args)
+    {
```
Comment (wmdietl) [inline checker/examples/units-extension/Demo.java]: Fix linebreak.

File: checker/examples/units-extension/Expected.txt [hunk 13]
```diff
@@ -0,0 +1,6 @@
+Demo.java:10: error: [assignment.type.incompatible] incompatible types in assignment.
+        frq = 5;
+              ^
+  found   : int
+  required: @Hz (@Hz :: int)
```
Comment (wmdietl) [inline checker/examples/units-extension/Expected.txt]: The output in "required" is not what a human would expect.
Is this also the output for errors when used on the command line?
Shouldn't this be "@Hz int"?

File: checker/examples/units-extension/Frequency.java [hunk 14]
```diff
@@ -6,6 +6,8 @@
 
 /**
  * Units of frequency, such as hertz (@{@link Hz}).
+ * Skip test below ensures the framework's automated tests does not consider this file as a test file.
```
Comment (wmdietl) [inline checker/examples/units-extension/Frequency.java]: Why is this necessary? This is just an annotation, so I wouldn't expect an error in this file.

File: checker/examples/units-extension/FrequencyRelations.java [hunk 15]
```diff
@@ -1,16 +1,22 @@
...
+import org.checkerframework.javacutil.AnnotationUtils;
 
 /*>>>
 import org.checkerframework.checker.nullness.qual.Nullable;
-*/
+ */
 
-/** Relations among units of frequency. */
+/** Relations among units of frequency.
+ * Skip test below ensures the framework's automated tests does not consider this file as a test file.
```
Comment (wmdietl) [inline checker/examples/units-extension/FrequencyRelations.java]: I again don't see why this file needs to be skipped. Should it be SupressWarnings for unwanted units warnings? Why are there units warnings?

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java [hunk 16]
```diff
@@ -99,17 +99,17 @@ public AnnotationMirror aliasedAnnotation(AnnotationMirror a) {
...
                     final Class<? extends Annotation> q =
                             (Class<? extends Annotation>) Class.forName(qualName);
-
+                    // add the annotation class to the qualifier set
                     qualSet.add(q);
+                    // process the units annotation and add its corresponding units relations class
                     addUnitsRelations(q);
                 } catch (ClassNotFoundException e) {
-                    checker.message(javax.tools.Diagnostic.Kind.WARNING,
+                    checker.message(javax.tools.Diagnostic.Kind.WARNING,
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: Remove trailing whitespace.

File: checker/examples/units-extension/FrequencyRelations.java [hunk 17]
```diff
@@ -35,11 +41,17 @@ public UnitsRelations init(ProcessingEnvironment env) {
...
             return hz;
         }
 
         return null;
     }
 
-}
+    // returns true if the type mirror either has no annotations or has only 1 annotation and it is UnknownUnits
+    // any number that isn't assigned a unit will automatically gets the annotation UnknownUnits, eg: int x = 5; //x has @UnknownUnits
+    private boolean noUnits(AnnotatedTypeMirror t) {
```
Comment (wmdietl) [inline checker/examples/units-extension/FrequencyRelations.java]: We should not put this in FrequencyRelations, as it gives a bad example and will result in all extensions copying the same code.
Can you put this helper method as a static method in UnitsRelationsDefault?
Then also UnitsAnnotatedTypeFactory can use that method instead of containing a copy.

File: checker/build.xml (hunk 11 above)
Comment (jyluo) [inline checker/build.xml]: Tainting and Regex have underscores in their targets:
regex-tests
regex_classic-tests
regex_qual-tests
tainting-tests
tainting_classic-tests
tainting_qual-tests

Should I fix those up as well?

File: checker/examples/units-extension/Expected.txt (hunk 13 above)
Comment (jyluo) [inline checker/examples/units-extension/Expected.txt]: This is the exact output on command line.

For the shipped units, we have the output format:
Multiples.java:109: error: [assignment.type.incompatible] incompatible types in assignment.
        @m int distance = s * s * accelm;
                                ^
  found   : @mPERs2 int
  required: @m (@org.checkerframework.checker.units.qual.m :: int)

The inner @Hz is simply due to the fact that the unit is not part of any package in the demo.

File: checker/examples/units-extension/Frequency.java (hunk 14 above)
Comment (jyluo) [inline checker/examples/units-extension/Frequency.java]: Removing in the next commit.

File: checker/build.xml (hunk 11 above)
Comment (mernst) [inline checker/build.xml]: I'm guessing the underscores were an attempt to indicate the parsing.
I don't have a problem with underscores.  But if you are going to remove some, then yes, please remove them all.  Thanks!

Comment (jyluo) [inline checker/build.xml]: Done!

**Comment by jyluo** [comment]
All code reviews are addressed in commit 5df3816

File: checker/examples/units-extension/Demo.java [hunk 18]
```diff
@@ -29,4 +37,26 @@ void auto(@s int time) {
...
+        @Hz int resultHz = hertz + 20 / seconds;
+        System.out.println(resultHz);
+
+        @kHz int resultkHz = kilohertz + 50 / millisec;
+        System.out.println(resultkHz);
+
+        // this demonstrates the type hierarchy resolution: the common supertype of Hz and kHz is Frequency, so this statement will pass
+        @Frequency int okTernaryAssign = seconds > 10 ? hertz : kilohertz;
+
+        // on the other hand, this statement expects the right hand side to be a Hz, so it will fail
```
Comment (wmdietl) [inline checker/examples/units-extension/Demo.java]: Why is there no //:: error if this line is expected to fail?

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java [hunk 19]
```diff
@@ -181,6 +208,93 @@ public AnnotationMirror aliasedAnnotation(AnnotationMirror a) {
...
+        // load external units
+        String qualNames = checker.getOption("units");
+        if (qualNames != null) {
+            for (String qualName : qualNames.split(",")) {
+                loadExternalUnit(qualName);
+            }
+        }
+    }
+
+    @SuppressWarnings("unchecked")
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: Can this be moved to a smaller context, e.g. a local variable?

File: checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java [hunk 20]
```diff
@@ -181,6 +208,93 @@ public AnnotationMirror aliasedAnnotation(AnnotationMirror a) {
...
+                // TODO: does every alias have to have Prefix?
+                return true;
+            }
+        }
+
+        // if we are unable to find UnitsMultiple meta annotation, then this is not an Aliased Annotation
+        return false;
+    }
+
+    @SuppressWarnings("unchecked")
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java]: Same as above, move the SuppressWarnings to the smallest possible scope.

File: checker/examples/units-extension/Demo.java (hunk 18 above)
Comment (jyluo) [inline checker/examples/units-extension/Demo.java]: The output is checked against Expected.txt as part of the make script. I can add //::error for documentation but this test isn't executed through the normal test framework due to the need to load the external units.

File: framework/src/org/checkerframework/framework/type/DefaultAnnotatedTypeFormatter.java [hunk 21]
```diff
@@ -355,7 +359,7 @@ public String visitWildcard(AnnotatedWildcardType type, Set<AnnotatedTypeMirror>
 
         @SideEffectFree
         protected String formatFlatType(final AnnotatedTypeMirror flatType) {
-            return annoFormatter.formatAnnotationString(flatType.getAnnotations(), currentPrintInvisibleSetting) + flatType.actualType;
+            return annoFormatter.formatAnnotationString(flatType.getAnnotations(), currentPrintInvisibleSetting) + TypeAnnotationUtils.unannotatedType((Type) flatType.getUnderlyingType());
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/DefaultAnnotatedTypeFormatter.java]: Break this up nicer over two lines.

File: checker/examples/units-extension/Demo.java (hunk 18 above)
Comment (wmdietl) [inline checker/examples/units-extension/Demo.java]: You should remove the //:: error comment on line 9 then.
